backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/ir/interop/cenum/CEnumClassGenerator.kt

```diff
@@ -97,38 +99,46 @@ internal class CEnumClassGenerator(
                     SYNTHETIC_OFFSET, SYNTHETIC_OFFSET, IrDeclarationOrigin.PROPERTY_BACKING_FIELD,
                     propertyDescriptor, propertyDescriptor.type.toIrType(), DescriptorVisibilities.PRIVATE
             ).also {
-                it.initializer = irBuilder(irBuiltIns, it.symbol, SYNTHETIC_OFFSET, SYNTHETIC_OFFSET).run {
-                    irExprBody(irGet(irClass.primaryConstructor!!.valueParameters[0]))
+                postLinkageSteps.add {
+                    it.initializer = irBuilder(irBuiltIns, it.symbol, SYNTHETIC_OFFSET, SYNTHETIC_OFFSET).run {
+                        irExprBody(irGet(irClass.primaryConstructor!!.valueParameters[0]))
+                    }
                 }
             }
         }
         val getter = irProperty.getter!!
         getter.correspondingPropertySymbol = irProperty.symbol
-        getter.body = irBuilder(irBuiltIns, getter.symbol, SYNTHETIC_OFFSET, SYNTHETIC_OFFSET).irBlockBody {
-            +irReturn(
-                    irGetField(
-                            irGet(getter.dispatchReceiverParameter!!),
-                            irProperty.backingField!!
-                    )
-            )
+        postLinkageSteps.add {
+            getter.body = irBuilder(irBuiltIns, getter.symbol, SYNTHETIC_OFFSET, SYNTHETIC_OFFSET).irBlockBody {
+                +irReturn(
+                        irGetField(
+                                irGet(getter.dispatchReceiverParameter!!),
+                                irProperty.backingField!!
+                        )
+                )
+            }
         }
         return irProperty
     }
 
     private fun createEnumEntry(enumDescriptor: ClassDescriptor, entryDescriptor: ClassDescriptor): IrEnumEntry {
-        return symbolTable.declareEnumEntry(
+        val enumEntry = symbolTable.declareEnumEntry(
                 SYNTHETIC_OFFSET, SYNTHETIC_OFFSET,
                 IrDeclarationOrigin.IR_EXTERNAL_DECLARATION_STUB, entryDescriptor
-        ).also { enumEntry ->
-            enumEntry.initializerExpression = IrExpressionBodyImpl(IrEnumConstructorCallImpl.fromSymbolDescriptor(
+        )
+        val constructorSymbol = symbolTable.referenceConstructor(enumDescriptor.unsubstitutedPrimaryConstructor!!)
+        postLinkageSteps.add {
+            enumEntry.initializerExpression = IrExpressionBodyImpl(IrEnumConstructorCallImpl(
                     SYNTHETIC_OFFSET, SYNTHETIC_OFFSET,
                     type = irBuiltIns.unitType,
-                    symbol = symbolTable.referenceConstructor(enumDescriptor.unsubstitutedPrimaryConstructor!!),
-                    typeArgumentsCount = 0 // enums can't be generic
+                    symbol = constructorSymbol,
+                    typeArgumentsCount = constructorSymbol.owner.typeParameters.size,
```

Comment:
Constructors can't have any type parameters unless they are declared in Java which is impossible on Native.